includes/Specials/SpecialDataDump.php

```diff
@@ -79,38 +83,84 @@ public function execute( $par ) {
 			}
 		}
 
-		$pager = new DataDumpPager( $this->getContext(), $this->getPageTitle() );
+		$pager = new DataDumpPager(
+			$this->getConfig(),
+			$this->getContext(),
+			$this->connectionProvider,
+			$this->jobQueueGroupFactory,
+			$this->getLinkRenderer(),
+			$this->permissionManager
+		);
 
 		$out->addModuleStyles( [ 'mediawiki.special' ] );
 
 		$pager->getForm();
 		$out->addParserOutputContent( $pager->getFullOutput() );
 	}
 
-	private function doDownload( string $fileName ) {
+	private function doDownload( string $fileName ): bool {
 		$out = $this->getOutput();
 		$out->disable();
 
 		$backend = DataDump::getBackend();
-		$backend->streamFile( [
-			'src'     =>
-				$backend->getContainerStoragePath( 'dumps-backup' ) . '/' . $fileName,
-			'headers' => [
-				'Expires: ' . gmdate( 'D, d M Y H:i:s', 0 ) . ' GMT',
-				'Cache-Control: no-cache, no-store, max-age=0, must-revalidate',
-				'Pragma: no-cache',
-				'Content-Disposition: attachment; filename="' . $fileName . '"',
-			]
-		] )->isOK();
+		$directoryBackend = $backend->getContainerStoragePath( 'dumps-backup' );
+
+		if ( $backend->fileExists( [ 'src' => $directoryBackend . '/' . $fileName ] ) ) {
+			$backend->streamFile( [
+				'src' => $directoryBackend . '/' . $fileName,
+				'headers' => [
+					'Expires: ' . gmdate( 'D, d M Y H:i:s', 0 ) . ' GMT',
+					'Cache-Control: no-cache, no-store, max-age=0, must-revalidate',
+					'Pragma: no-cache',
+					'Content-Disposition: attachment; filename="' . $fileName . '"',
+				]
+			] )->isOK();
+		} else {
+			$this->streamFileChunks( $fileName, $directoryBackend, $backend );
+		}
 
 		return true;
 	}
 
-	private function doDelete( string $type, string $fileName ) {
-		$dataDumpConfig = $this->config->get( 'DataDump' );
+	private function streamFileChunks(
+		string $fileName,
+		string $directoryBackend,
+		FileBackend $backend
+	): void {
+		$chunkIndex = 0;
+		$chunkFileName = $fileName . '.part' . $chunkIndex;
+		$headersSent = false;
+
+		while ( $backend->fileExists( [ 'src' => $directoryBackend . '/' . $chunkFileName ] ) ) {
+			if ( !$headersSent ) {
+				header( 'Expires: ' . gmdate( 'D, d M Y H:i:s', 0 ) . ' GMT' );
```

Comment:
I wonder if we should hardcode the timestamp here

Reply:
What do you mean?